revoke/revoke.go

```diff
@@ -22,14 +24,85 @@ import (
 	"github.com/cloudflare/cfssl/log"
 )
 
-// HardFail determines whether the failure to check the revocation
-// status of a certificate (i.e. due to network failure) causes
-// verification to fail (a hard failure).
-var HardFail = false
+// Revoke type contains configuration for each new revoke instance
+type Revoke struct {
+	// lock is a Mutex locker for the struct
+	lock sync.Mutex
+	// LocalCRL contains path to the local CRL file. When set, certificate
+	// will be checked only using local CRL script, remote methods will be
+	// skipped.
+	localCRL string
+	// HardFail determines whether the failure to check the revocation
+	// status of a certificate (i.e. due to network failure) causes
+	// verification to fail (a hard failure).
+	hardFail bool
+	// crlSet associates a PKIX certificate list with the URL the CRL is
+	// fetched from.
+	crlSet map[string]*pkix.CertificateList
+}
+
+// defaultChecker is a default config for regular apps which don't need to
+// use custom options.
+var defaultChecker = New(false)
+
+// New creates Revoke config structure.
+// Accepts hardfail bool variable as an option
+func New(hardfail bool) *Revoke {
+	return &Revoke{
+		localCRL: "",
+		hardFail: hardfail,
+		crlSet:   map[string]*pkix.CertificateList{},
+	}
+}
+
+// SetLocalCRL sets localCRL path into the Revoke struct
+func (r *Revoke) SetLocalCRL(localCRLpath string) error {
+	if localCRLpath == "" {
+		r.lock.Lock()
```

Comment:
The practice of lock/unlock tightly around a read/write operation isn't going to buy something. In fact this introduces the exact race condition that deletion is reaching lock faster than write, so a cache deletion after a cache write may have chances finishing before write finishes, making deletion an no-op. My suggestion is as follows:
1. lock at the very beginning of a read/write/deletion function. defer unlock. Let's not worry about efficiency right now, correctness must be maintained as the first priority.
2. the CRL fetch works as follows: 1. check hashmap, return hit. 2. in case of miss, fetch from remote and update hashmap
3. revocation check actually happens after crl fetch, so revocation check is lock free.